File: src/common/web/api/user/UserEvents.ts
import { Type } from "class-transformer";

import { EventComposer } from "../../core/messaging/composers/EventComposer";
import { MessageBuilder } from "../../core/messaging/composers/MessageBuilder";
import { Event } from "../../core/messaging/Event";
import { Message } from "../../core/messaging/Message";

/**
 * Event sent when the user authorizations have changed.
 *
 * @param authorizations - The new list of all granted authorizations.
 */
@Message.define("event/user/authorization/list")
export class UserAuthorizationsListEvent extends Event {
    // @ts-ignore
    @Type(() => String)
    public readonly authorizations: string[] = [];

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, authorizations: string[], chain: Message | null = null): EventComposer<UserAuthorizationsListEvent> {
        return messageBuilder.buildEvent(UserAuthorizationsListEvent, { authorizations: authorizations }, chain);
    }
}

File: src/common/py/api/user/user_events.py
import dataclasses
import typing

from ...core.messaging import (
    Event,
    Message,
)
from ...core.messaging.composers import (
    EventComposer,
    MessageBuilder,
)


@Message.define("event/user/authorization/list")
class UserAuthorizationsListEvent(Event):
    """
    Event sent when the user authorizations have changed.

    Args:
        authorizations: The new list of all granted authorizations.
    """

    authorizations: typing.List[str] = dataclasses.field(default_factory=list)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        authorizations: typing.List[str],
        chain: Message | None = None,
    ) -> EventComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_event(
            UserAuthorizationsListEvent,
            chain,
            authorizations=authorizations,
        )
